### PR TITLE
Unshorten url

### PR DESCRIPTION
**Pull request by antonydp**
Utils class to bypass shortner URLs
Based on https://github.com/kodiondemand/addon/blob/master/lib/unshortenit.py
Still a work in progress.
Since there's still a lot of work to be done, I would like to hear some opinions as to whether this can be useful to the project

### CODE REVIEW

**Comment by antonydp** [comment]
Since there's still a lot of work to be done, I would like to hear some opinions if this could be helpful to the project

**Comment by ArjixWasTaken** [comment]
In [this](https://github.com/LagradOst/CloudStream-3/pull/1172/files#diff-363538209598c6be246664b79351cf92a714539045615befb5670b37fe28e4ebR37-R118) function please use an `if else` chain

Like
```kotlin
if (false) { /* do smth */ }
else if (true) { /* do smth else */ }
```

**Comment by ArjixWasTaken** [comment]
And please use `camelCase` instead of `snake_case`.

**Comment by ArjixWasTaken** [comment]
> Since there's still a lot of work to be done, I would like to hear some opinions as to whether this can be useful to the project

Well, if it is guaranteed to work, then I bet it will be useful, in the past when we needed to bypass a shrinked url we just implemented the bypass on a provider basis.

For example [here](https://github.com/LagradOst/CloudStream-3/blob/master/app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt#L385-L451) we bypass adfly.
Now that bypass may be outdated, who knows, the provider has been dead for like a year now, so it is unused.

**Comment by antonydp** [comment]
I've rewritten the snake_case in camelCase as suggested and used a loop over a list instead of the if chain.
I've removed the not yet coded shortner, witch can be found here in the repository i've been taking inspiration: 
https://github.com/kodiondemand/addon/blob/master/lib/unshortenit.py 
I think it can be merged right now as we can keep adding useful url shortner as we go.

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    var currentUrl = uri
+    val adflyRegex = """adf\.ly|j\.gs|q\.gs|u\.bb|ay\.gy|atominik\.com|tinyium\.com|microify\.com|threadsphere\.bid|clearload\.bid|activetect\.net|swiftviz\.net|briskgram\.net|activetect\.net|stoodsef\.com|baymaleti\.net|thouth\.net|uclaut\.net|gloyah\.net|larati\.net|scuseami\.net"""
+    val linkupRegex = """linkup\.pro|buckler.link"""
+    val linksafeRegex = """linksafe\.cc"""
+    val nuovoIndirizzoRegex = """mixdrop\.nuovoindirizzo\.com"""
+    val nuovoLinkRegex = """nuovolink\.com"""
+    val uprotRegex = """uprot\.net"""
+    val davisonbarkerRegex = """davisonbarker\.pro|lowrihouston\.pro"""
+
+    val shortList = listOf(
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:32]: make this a companion object as well as the regex links, as you dont want to run unnecessary code for every unshorten.

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    when{
+        uri.contains("/tv/") -> uri = uri.replace("/tv/", "/tva/")
+        uri.contains("delta") -> uri = uri.replace("/delta/", "/adelta/")
+        (uri.contains("/ga/") || uri.contains("/ga2/")) -> uri = base64Decode(uri.split('/').last()).trim()
+        uri.contains("/speedx/") -> uri = uri.replace("http://linkup.pro/speedx", "http://speedvideo.net")
+        else -> {
+            r = app.get(uri, allowRedirects = true)
+            uri = r.url
+            val link =
+                Regex("<iframe[^<>]*src=\\'([^'>]*)\\'[^<>]*>").find(r.text)?.value ?:
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:101]: the code will be slightly faster if you move regex to companion

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
+package com.lagradost.cloudstream3.utils
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:1]: impo this should be an object and not just code

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            uri = Regex("url=(.*)\$").find(uri)!!.value
+        }
+        else if (uri.contains("/decode/")) {
+            uri = app.get(uri, allowRedirects = true).url
+        }
+    }
+    return uri
+}
+fun unshortenLinksafe(uri:String) : String {
+    return Base64.decode(uri.split("?url=").last().toByteArray(), Base64.DEFAULT).decodeToString()
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:134]: use mainapis base64 decode

**Review by LagradOst — COMMENTED** [review]
👍 just minor code changes, mainly that it should be an object and stuff should be computed once (like regex and shortList)

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    val type: String,
+    val function: suspend (String) -> String,
+) {
+    constructor(regex: String, type: String, function: suspend (String) -> String) : this(
+        Regex(regex),
+        type,
+        function
+    )
+}
+suspend fun unshorten(uri: String, type: String?): String {
```
[the remark this answers is not in the document]
Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt]: yes!

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt [hunk 6]
```diff
@@ -0,0 +1,161 @@
+package com.lagradost.cloudstream3.utils
+import android.util.Base64
+import com.lagradost.cloudstream3.app
+import com.lagradost.cloudstream3.base64Decode
+import com.lagradost.nicehttp.NiceResponse
+import java.net.URI
+import java.net.URLDecoder
+
+//Code heavily based on unshortenit.py form kodiondemand /addon
```
[the remark this answers is not in the document]
Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:9]: I think I've done it.

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt (hunk 1 above)
Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt]: is it ok the way I did it?

Comment (LagradOst) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt]: yes

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt (hunk 2 above)
Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:110]: I don't really understand what should I do, sorry

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt (hunk 3 above)
Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:1]: that's not entirely clear to me, sorry. could you explain a little bit more?

Comment (LagradOst) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:1]:
```kotlin
object ShortUrl { 
[code here]
}
```
so you have to prefix it with ShortUrl.isShortLink when calling it

Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:1]: for both isShortLink and unshorten?

Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:1]: I wrote the ShortLink class, is it ok like that?

File: app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt (hunk 4 above)
Comment (antonydp) [reply app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt]: done!

**Comment by LagradOst** [comment]
Very nice, do you want this merged now or do you want to add smth more?

**Comment by antonydp** [comment]
I think it can be merged now, I'll try to add some more shorteners later on

**Pull request merged**